client-api/src/main/java/org/sonarsource/sonarlint/core/client/api/util/FileUtils.java

```diff
@@ -39,13 +42,58 @@ private FileUtils() {
     // utility class, forbidden constructor
   }
 
-  public static void moveDir(Path temp, Path dest) {
+  public static void moveDir(Path src, Path dest) {
     try {
-      Files.move(temp, dest, StandardCopyOption.ATOMIC_MOVE);
-    } catch (AtomicMoveNotSupportedException e) {
-      throw new IllegalStateException("Atomic move not supported for " + temp + " to " + dest, e);
+      moveDirPreferAtomic(src, dest);
     } catch (IOException e) {
-      throw new IllegalStateException("Unable to move " + temp + " to " + dest, e);
+      throw new IllegalStateException("Unable to move " + src + " to " + dest, e);
+    }
+  }
+
+  private static void moveDirPreferAtomic(Path src, Path dest) throws IOException {
+    try {
+      Files.move(src, dest, StandardCopyOption.ATOMIC_MOVE);
+    } catch (AtomicMoveNotSupportedException e) {
+      // Fallback to non atomic move
+      Files.walkFileTree(src, new MoveRecursivelyVisitor(src, dest));
+    }
+  }
+
+  private static class MoveRecursivelyVisitor extends SimpleFileVisitor<Path> {
+    private final Path fromPath;
+    private final Path toPath;
+    private final CopyOption[] copyOptions;
+
+    public MoveRecursivelyVisitor(Path fromPath, Path toPath, CopyOption... copyOptions) {
+      this.fromPath = fromPath;
+      this.toPath = toPath;
+      this.copyOptions = copyOptions;
+    }
+
+    @Override
+    public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+      Path targetPath = toPath.resolve(fromPath.relativize(dir));
+      if (!Files.exists(targetPath)) {
+        Files.createDirectory(targetPath);
+      }
+      return FileVisitResult.CONTINUE;
+    }
+
+    @Override
+    public FileVisitResult postVisitDirectory(Path dir, @Nullable IOException exc) throws IOException {
+      Objects.requireNonNull(dir);
+      if (exc != null) {
+        throw exc;
+      }
+      // Dir should now be empty
+      Files.delete(dir);
```

Comment:
If a problem occurs during the directory tree move, we will end up with a partial directory at the destination and at the source. It would be better to make a complete copy, and then delete the source. That way at least one location will be guaranteed to have the complete content.